File: vortex/src/vortex/dataset.py
"""Dataset registration and management."""

from dataclasses import dataclass
from typing import List, Optional, Iterator
import logging

from vortex.config import DatasetConfig, ShardAssignment

logger = logging.getLogger(__name__)


@dataclass
class Dataset:
    """A registered dataset with shard assignments.
    
    This class represents a dataset that has been registered with the
    coordinator and has shards assigned to this worker.
    """
    id: str
    config: DatasetConfig
    assigned_shards: List[int]
    _current_idx: int = 0
    
    def __iter__(self) -> Iterator[int]:
        """Iterate over assigned shard IDs."""
        return iter(self.assigned_shards)
    
    def __len__(self) -> int:
        """Number of assigned shards."""
        return len(self.assigned_shards)
    
    def next_shard(self) -> Optional[int]:
        """Get the next shard ID to process."""
        if self._current_idx >= len(self.assigned_shards):
            return None
        shard_id = self.assigned_shards[self._current_idx]
        self._current_idx += 1
        return shard_id
    
    def reset(self) -> None:
        """Reset to the beginning of assigned shards."""
        self._current_idx = 0
    
    def is_exhausted(self) -> bool:
        """Check if all shards have been consumed."""
        return self._current_idx >= len(self.assigned_shards)
    
    def shard_key(self, shard_id: int) -> str:
        """Get the storage key for a shard."""
        pattern = self.config.shard_pattern or "shard_{:06d}.bin"
        filename = pattern.format(shard_id)
        return f"{self.config.shard_prefix}{filename}"


# Global dataset registry
_datasets: dict[str, Dataset] = {}


def register_dataset(
    name: str,
    shard_prefix: str,
    num_shards: int,
    shuffle_shards: bool = True,
    seed: int = 42,
    shard_pattern: Optional[str] = None,
) -> Dataset:
    """Register a dataset with the coordinator.
    
    This function registers a dataset and requests shard assignments
    from the coordinator. The registration is idempotent - calling
    with the same name will return the existing dataset.
    
    Args:
        name: Unique dataset identifier
        shard_prefix: S3 prefix for shards
        num_shards: Total number of shards
        shuffle_shards: Whether to shuffle shards
        seed: Random seed for shuffling
        shard_pattern: Optional filename pattern
        
    Returns:
        Dataset: The registered dataset with shard assignments
    """
    # Check if already registered
    if name in _datasets:
        logger.info(f"Dataset '{name}' already registered, returning existing")
        return _datasets[name]
    
    config = DatasetConfig(
        name=name,
        shard_prefix=shard_prefix,
        num_shards=num_shards,
        shuffle_shards=shuffle_shards,
        seed=seed,
        shard_pattern=shard_pattern,
    )
    
    # In real implementation, this calls into Rust via PyO3
    # to register with coordinator and get shard assignments
    # For now, simulate with placeholder
    logger.info(f"Registering dataset '{name}' with {num_shards} shards")
    
    # Placeholder: would come from coordinator
    assigned_shards = list(range(num_shards))  # Worker 0 gets all in single-worker mode
    
    dataset = Dataset(
        id=name,
        config=config,
        assigned_shards=assigned_shards,
    )
    
    _datasets[name] = dataset
    return dataset


def get_dataset(name: str) -> Optional[Dataset]:
    """Get a registered dataset by name."""
    return _datasets.get(name)


def list_datasets() -> List[str]:
    """List all registered dataset names."""
    return list(_datasets.keys())

File: vortex-core/src/data/dataset.rs
//! Dataset abstraction and registration
//!
//! Provides dataset metadata and configuration for sharded datasets.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Dataset configuration for registration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatasetConfig {
    /// Unique dataset identifier
    pub name: String,
    /// S3 prefix for shards (e.g., "s3://bucket/dataset/shards/")
    pub shard_prefix: String,
    /// Total number of shards
    pub num_shards: u32,
    /// Whether to shuffle shards across workers
    pub shuffle_shards: bool,
    /// Random seed for shuffling
    pub seed: u64,
    /// Optional shard file pattern (default: "shard_{:06d}.bin")
    pub shard_pattern: Option<String>,
}

/// Dataset metadata stored in object storage
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatasetMetadata {
    /// Dataset configuration
    pub config: DatasetConfig,
    /// Creation timestamp
    pub created_at: chrono::DateTime<chrono::Utc>,
    /// Total size in bytes (if known)
    pub total_bytes: Option<u64>,
    /// Per-shard metadata
    pub shard_metadata: HashMap<u32, ShardMetadata>,
}

/// Metadata for a single shard
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShardMetadata {
    /// Shard ID
    pub shard_id: u32,
    /// Object key in storage
    pub object_key: String,
    /// Size in bytes
    pub byte_size: u64,
    /// Number of records (if known)
    pub record_count: Option<u64>,
    /// CRC32C checksum
    pub crc32c: Option<u32>,
}

/// Runtime representation of a dataset
#[derive(Debug, Clone)]
pub struct Dataset {
    /// Unique identifier
    pub id: String,
    /// Dataset metadata
    pub metadata: DatasetMetadata,
    /// Shard assignment for this worker
    pub assigned_shards: Vec<u32>,
    /// Current shard index
    current_shard_idx: usize,
}

impl Dataset {
    /// Create a new dataset from metadata
    pub fn new(id: String, metadata: DatasetMetadata) -> Self {
        Self {
            id,
            metadata,
            assigned_shards: Vec::new(),
            current_shard_idx: 0,
        }
    }

    /// Set the assigned shards for this worker
    pub fn set_assigned_shards(&mut self, shards: Vec<u32>) {
        self.assigned_shards = shards;
        self.current_shard_idx = 0;
    }

    /// Get the next shard to load
    pub fn next_shard(&mut self) -> Option<u32> {
        if self.current_shard_idx >= self.assigned_shards.len() {
            return None;
        }
        let shard = self.assigned_shards[self.current_shard_idx];
        self.current_shard_idx += 1;
        Some(shard)
    }

    /// Reset to the beginning of assigned shards
    pub fn reset(&mut self) {
        self.current_shard_idx = 0;
    }

    /// Check if all shards have been consumed
    pub fn is_exhausted(&self) -> bool {
        self.current_shard_idx >= self.assigned_shards.len()
    }

    /// Get shard object key
    pub fn shard_key(&self, shard_id: u32) -> String {
        let pattern = self.metadata.config.shard_pattern
            .as_deref()
            .unwrap_or("shard_{:06d}.bin");
        
        let filename = pattern.replace("{:06d}", &format!("{:06}", shard_id));
        format!("{}{}", self.metadata.config.shard_prefix, filename)
    }
}

impl DatasetConfig {
    /// Create a new dataset configuration
    pub fn new(name: impl Into<String>, shard_prefix: impl Into<String>, num_shards: u32) -> Self {
        Self {
            name: name.into(),
            shard_prefix: shard_prefix.into(),
            num_shards,
            shuffle_shards: true,
            seed: 42,
            shard_pattern: None,
        }
    }
}
